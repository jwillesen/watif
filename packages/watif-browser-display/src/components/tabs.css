.tabs {
  display: flex;
  justify-content: center;
}

.tab {
  padding: 0.5rem;
  border: var(--tab-border-width) solid var(--tab-background-color);
  background: var(--tab-background-color);
  flex: 1;
}

.selected {
  background: var(--tab-selected-background-color);
}