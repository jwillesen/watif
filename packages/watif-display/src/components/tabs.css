.tab-container {
  display: flex;
  flex-direction: column;
  height: 100%;
}

.tab-bar {
  display: flex;
  background: var(--tab-background-color);
  border-radius: 1rem;
}

.tab-button {
  padding: 0.5rem;
  border: var(--tab-border-width) solid var(--tab-background-color);
  border-radius: 1rem;
  background: var(--tab-background-color);
  flex: 1;
}

.tab-panels {
  flex: 1;
}

.tab-panel {
  height: 100%;
}

.selected {
  background: var(--tab-selected-background-color);
}
