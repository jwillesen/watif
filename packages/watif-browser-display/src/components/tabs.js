import React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faBook, faEye, faBox, faHandHoldingBox } from '@fortawesome/pro-regular-svg-icons'

import './tabs.css'

export default class Tabs extends React.Component {
  constructor (...args) {
    super(...args)
    this.state = {
      activeTabLabel: 'Events',
    }
  }

  selectEvents = () => this.setState({activeTabLabel: 'Events'})
  selectLook = () => this.setState({activeTabLabel: 'Look Around'})
  selectExamine = () => this.setState({activeTabLabel: 'Examine'})
  selectInventory = () => this.setState({activeTabLabel: 'Inventory'})

  renderTab (label, icon, onClick) {
    const styles = ['tab']
    if (label === this.state.activeTabLabel) styles.push('selected')

    return <button
      styleName={styles.join(' ')}
      role="tab"
      aria-label={label}
      onClick={onClick}
    >
      <FontAwesomeIcon icon={icon} size="lg" />
    </button>
  }

  render () {
    return <div
      styleName="tabs"
      role="tablist"
      aria-label="Events, Look Around, Examine, and Inventory"
    >
      {this.renderTab('Events', faBook, this.selectEvents)}
      {this.renderTab('Look Around', faEye, this.selectLook)}
      {this.renderTab('Examine', faBox, this.selectExamine)}
      {this.renderTab('Inventory', faHandHoldingBox, this.selectInventory)}
    </div>
  }
}
