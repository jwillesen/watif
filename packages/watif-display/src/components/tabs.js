import React from 'react'
import {arrayOf, oneOfType, element, string, func} from 'prop-types'

import './tabs.css'

export class Tabs extends React.Component {
  buttonElts = {}

  static propTypes = {
    label: string.isRequired,
    children: oneOfType([arrayOf(element), element]), // TabPanel elements
    activeTab: string,
    onTabRequest: func, // (TabPanel.props.id of requested tab)
  }

  componentDidUpdate(prevProps) {
    if (prevProps.activeTab !== this.props.activeTab) {
      const elt = this.buttonElts[this.props.activeTab]
      if (elt) elt.focus()
    }
  }

  tabButtonId(panel) {
    return `tab-button-${panel.props.id}`
  }

  tabPanelId(panel) {
    return `tab-panel-${panel.props.id}`
  }

  onKey = event => {
    const panels = React.Children.toArray(this.props.children)
    const activeIndex = panels.findIndex(panel => panel.props.id === this.props.activeTab)
    let newIndex
    if (event.key === 'ArrowRight') {
      newIndex = activeIndex + 1
      if (newIndex >= panels.length) newIndex = 0
    } else if (event.key === 'ArrowLeft') {
      newIndex = activeIndex - 1
      if (newIndex < 0) newIndex = panels.length - 1
    } else if (event.key === 'Home') {
      newIndex = 0
    } else if (event.key === 'End') {
      newIndex = panels.length - 1
    }

    if (newIndex !== undefined) {
      const newPanel = panels[newIndex]
      this.props.onTabRequest(newPanel.props.id)
    }
  }

  renderButton(panel) {
    const styles = ['tab-button']
    let tabIndex = -1
    if (panel.props.id === this.props.activeTab) {
      styles.push('selected')
      tabIndex = 0
    }

    return (
      <button
        type="button"
        key={panel.props.id}
        ref={elt => (this.buttonElts[panel.props.id] = elt)}
        styleName={styles.join(' ')}
        id={this.tabButtonId(panel)}
        role="tab"
        tabIndex={tabIndex}
        aria-label={panel.props.a11yLabel}
        aria-selected={tabIndex === 0}
        aria-controls={this.tabPanelId(panel)}
        onClick={() => this.props.onTabRequest(panel.props.id)}
        onKeyDown={this.onKey}
      >
        {panel.props.label}
      </button>
    )
  }

  renderPanel(panel) {
    const hidden = this.props.activeTab !== panel.props.id
    return (
      <div
        key={panel.props.id}
        styleName="tab-panel"
        id={this.tabPanelId(panel)}
        role="tabpanel"
        aria-labelledby={this.tabButtonId(panel)}
        tabIndex={0}
        hidden={hidden}
      >
        {panel.props.children}
      </div>
    )
  }

  render() {
    return (
      <div styleName="tab-container">
        <div styleName="tab-bar" role="tablist" aria-label={this.props.label}>
          {React.Children.map(this.props.children, panel => this.renderButton(panel))}
        </div>
        <div styleName="tab-panels">
          {React.Children.map(this.props.children, panel => this.renderPanel(panel))}
        </div>
      </div>
    )
  }
}

// This is just an interface component for documentation
/* eslint-disable react/no-multi-comp, react/no-unused-prop-types */
export class TabPanel extends React.Component {
  static propTypes = {
    children: oneOfType([arrayOf(element), element, string]),
    id: string,
    label: oneOfType([element, string]),
    a11yLabel: string,
  }

  render() {
    return this.props.children
  }
}
