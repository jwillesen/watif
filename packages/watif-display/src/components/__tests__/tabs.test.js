import React from 'react'
import {shallow, mount} from 'enzyme'

import { Tabs, TabPanel } from '../tabs'

function basicTabs () {
  return <Tabs label="some tabs">
    <TabPanel id="first" label="first tab" a11yLabel="first tab a11y">First</TabPanel>
    <TabPanel id="second" label="second tab" a11yLabel="second tab a11y">Second</TabPanel>
    <TabPanel id="third" label="third tab" a11yLabel="third tab a11y">Third</TabPanel>
    <TabPanel id="fourth" label="fourth tab" a11yLabel="fourth tab a11y">Fourth</TabPanel>
  </Tabs>
}

it('renders the things', () => {
  const wrapper = mount(basicTabs())
  expect(wrapper).toMatchSnapshot()
})

it('renders one selected tab', () => {
  const wrapper = shallow(basicTabs())
  expect(wrapper.find('button[role="tab"]')).toHaveLength(4)
  expect(wrapper.find('button[aria-selected=true]')).toHaveLength(1)
  expect(wrapper.find('button.selected').text()).toBe('first tab')
  expect(wrapper.find('button[aria-selected=false]')).toHaveLength(3)
})

it('manages tab index', () => {
  const wrapper = shallow(basicTabs())
  expect(wrapper.find('button[tabIndex=0]')).toHaveLength(1)
  expect(wrapper.find('button[tabIndex=-1]')).toHaveLength(3)
})

it('manages hidden', () => {
  const wrapper = shallow(basicTabs())
  expect(wrapper.find('div[hidden=false]')).toHaveLength(1)
  expect(wrapper.find('div[hidden=true]')).toHaveLength(3)
})

it('changes the selected tab on click', () => {
  const wrapper = shallow(basicTabs())
  wrapper.find('button[aria-selected=false]').first().simulate('click')
  const selectedButton = wrapper.find('button.selected')
  expect(selectedButton.text()).toEqual('second tab')
  expect(selectedButton.props().tabIndex).toBe(0)
  expect(wrapper.find('button[tabIndex=-1]')).toHaveLength(3)
})

it('changes the selected tab on right arrow', () => {
  const wrapper = mount(basicTabs())
  wrapper.find('button.selected').simulate('keyDown', {key: 'ArrowRight'})
  const selectedButton = wrapper.find('button.selected')
  expect(selectedButton.text()).toBe('second tab')
  expect(document.activeElement).toBe(selectedButton.instance())
})

it('wraps around to first tab on right arrow', () => {
  const wrapper = mount(basicTabs())
  wrapper.find('button').last().simulate('click')
  expect(wrapper.find('button.selected').text()).toBe('fourth tab')
  wrapper.find('button.selected').simulate('keyDown', {key: 'ArrowRight'})
  const selectedButton = wrapper.find('button.selected')
  expect(selectedButton.text()).toBe('first tab')
  expect(document.activeElement).toBe(selectedButton.instance())
})

it('changes the selected tab on left arrow', () => {
  const wrapper = mount(basicTabs())
  wrapper.find('button').last().simulate('click')
  expect(wrapper.find('button.selected').text()).toBe('fourth tab')
  wrapper.find('button.selected').simulate('keyDown', {key: 'ArrowLeft'})
  const selectedButton = wrapper.find('button.selected')
  expect(selectedButton.text()).toBe('third tab')
  expect(document.activeElement).toBe(selectedButton.instance())
})

it('wraps around to the last tab on left arrow', () => {
  const wrapper = mount(basicTabs())
  wrapper.find('button.selected').simulate('keyDown', {key: 'ArrowLeft'})
  const selectedButton = wrapper.find('button.selected')
  expect(selectedButton.text()).toBe('fourth tab')
  expect(document.activeElement).toBe(selectedButton.instance())
})

it('activates the first tab on the home', () => {
  const wrapper = mount(basicTabs())
  wrapper.find('button').last().simulate('click')
  wrapper.find('button.selected').simulate('keyDown', {key: 'Home'})
  const selectedButton = wrapper.find('button.selected')
  expect(selectedButton.text()).toBe('first tab')
  expect(document.activeElement).toBe(selectedButton.instance())
})

it('activates the first tab on the end', () => {
  const wrapper = mount(basicTabs())
  wrapper.find('button').last().simulate('click')
  wrapper.find('button.selected').simulate('keyDown', {key: 'End'})
  const selectedButton = wrapper.find('button.selected')
  expect(selectedButton.text()).toBe('fourth tab')
  expect(document.activeElement).toBe(selectedButton.instance())
})
