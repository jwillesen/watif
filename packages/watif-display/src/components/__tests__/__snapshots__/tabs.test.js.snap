// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the things 1`] = `
<Tabs
  label="some tabs"
>
  <div
    className="tab-container"
  >
    <div
      aria-label="some tabs"
      className="tab-bar"
      role="tablist"
    >
      <button
        aria-controls="tab-panel-first"
        aria-label="first tab a11y"
        aria-selected={true}
        className="tab-button selected"
        id="tab-button-first"
        key="first/.0"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={0}
      >
        first tab
      </button>
      <button
        aria-controls="tab-panel-second"
        aria-label="second tab a11y"
        aria-selected={false}
        className="tab-button"
        id="tab-button-second"
        key="second/.1"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={-1}
      >
        second tab
      </button>
      <button
        aria-controls="tab-panel-third"
        aria-label="third tab a11y"
        aria-selected={false}
        className="tab-button"
        id="tab-button-third"
        key="third/.2"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={-1}
      >
        third tab
      </button>
      <button
        aria-controls="tab-panel-fourth"
        aria-label="fourth tab a11y"
        aria-selected={false}
        className="tab-button"
        id="tab-button-fourth"
        key="fourth/.3"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={-1}
      >
        fourth tab
      </button>
    </div>
    <div
      className="tab-panels"
    >
      <div
        aria-labelledby="tab-button-first"
        className="tab-panel"
        hidden={false}
        id="tab-panel-first"
        key="first/.0"
        role="tabpanel"
        tabIndex={0}
      >
        First
      </div>
      <div
        aria-labelledby="tab-button-second"
        className="tab-panel"
        hidden={true}
        id="tab-panel-second"
        key="second/.1"
        role="tabpanel"
        tabIndex={0}
      >
        Second
      </div>
      <div
        aria-labelledby="tab-button-third"
        className="tab-panel"
        hidden={true}
        id="tab-panel-third"
        key="third/.2"
        role="tabpanel"
        tabIndex={0}
      >
        Third
      </div>
      <div
        aria-labelledby="tab-button-fourth"
        className="tab-panel"
        hidden={true}
        id="tab-panel-fourth"
        key="fourth/.3"
        role="tabpanel"
        tabIndex={0}
      >
        Fourth
      </div>
    </div>
  </div>
</Tabs>
`;
